### PR TITLE
ipc4: add mixin gain support

### PR DESCRIPTION
**Pull request by serhiy-katsyuba-intel**
Add possibility to apply gain to mixed sources to avoid
mixed output saturation. Gain (attenuation) is configurable
per source and comes from host via IPC4 large config set.

Signed-off-by: Serhiy Katsyuba <serhiy.katsyuba@intel.com>

### CODE REVIEW

File: src/audio/mixin_mixout.c [hunk 1]
```diff
@@ -262,113 +311,203 @@ static void audio_stream_bytes_copy(struct audio_stream __sparse_cache *dst_stre
...
+		while (bytes_to_copy != 0) {
+			n = MIN(audio_stream_bytes_without_wrap(sink, dest), bytes_to_copy);
+			n = MIN(audio_stream_bytes_without_wrap(source, src), n);
+			bytes_to_copy -= n;
+			n >>= 2;	/* divide by 4 */
+			for (i = 0; i < n; i++) {
+				int64_t s = (int64_t)(*src << 8);	/* extending sign bit */
+				s *= gain;
+				s >>= (GAIN_DIV_SHIFT + 8);
+				*dest = (int32_t)s;
```
Comment (marcinszkudlinski) [inline src/audio/mixin_mixout.c:445]: sat_int24?

Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:445]: Saturation check is not needed here as gain is always <=1.0 (1024 as integer). So the resulting value is equal or less then source value.

File: src/audio/mixin_mixout.c [hunk 2]
```diff
@@ -124,12 +129,50 @@ struct mixout_data {
...
+	uint32_t reserved2;
+	uint32_t reserved3;
+
+	/* Gain to be applied to input signal. Valid range: 0x0..0x400 (0.0 <= gain <= 1.0). Values
+	 * greater than 0x400 are treated as 0x400 (unity gain). To apply gain, multiply sample
+	 * by "gain" and divide by 1024.
+	 */
+	uint16_t gain;
+	uint16_t reserved4;
+} __packed __aligned(4);
```
Comment (lgirdwood) [inline src/audio/mixin_mixout.c:152]: Should these packed structures be in an IPC header ?

File: src/audio/mixin_mixout.c [hunk 3]
```diff
@@ -42,6 +42,8 @@ DECLARE_SOF_RT_UUID("mix_out", mixout_uuid, 0x3c56505a, 0x24d7, 0x418f,
 		    0xbd, 0xdc, 0xc1, 0xf5, 0xa3, 0xac, 0x2a, 0xe0);
 DECLARE_TR_CTX(mixout_tr, SOF_UUID(mixout_uuid), LOG_LEVEL_INFO);
 
+#define MIXIN_MAX_SINKS 3
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/mixin_mixout.c:45]: Fwiw, you could use Kconfig here to set this value if needed.

**Review by lgirdwood — COMMENTED** [review]
@serhiy-katsyuba-intel can we reuse the existing gain functions instead of adding more ? This would also buy us HiFi support.
@andrula-song fyi

File: src/audio/mixin_mixout.c (hunk 2 above)
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:152]: You are right. Will move then there.

Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Moved them to include/ipc4/mixin_mixout.h

**Comment by serhiy-katsyuba-intel** [comment]
> @serhiy-katsyuba-intel can we reuse the existing gain functions instead of adding more ? This would also buy us HiFi support.
@andrula-song fyi

Updated the PR to (1) use gain, saturation, extend sign math from format.h and (2) moved ipc4 specific structures to include/ipc4/mixin_mixout.h.

**Review by marcinszkudlinski — APPROVED** [review]

File: src/audio/mixin_mixout.c [hunk 4]
```diff
@@ -1020,6 +1111,73 @@ static int mixout_get_attribute(struct comp_dev *dev, uint32_t type, void *value
...
 }
 
+static int mixin_set_large_config(struct comp_dev *dev, uint32_t param_id, bool first_block,
+				  bool last_block, uint32_t data_offset_or_size, char *data)
+{
+	struct ipc4_mixer_mode_config *cfg;
+	struct mixin_data *mixin_data;
+	int i;
+	uint32_t sink_index;
+	uint16_t g;
```
Comment (lyakh) [inline src/audio/mixin_mixout.c:1121]: if you make a new iteration of this PR - maybe give this variable a longer name, e.g. `gain`. Short names are inconvenient to search for

File: src/audio/mixin_mixout.c [hunk 5]
```diff
@@ -1020,6 +1111,73 @@ static int mixout_get_attribute(struct comp_dev *dev, uint32_t type, void *value
...
+	}
+
+	/* for a single chunk data, data_offset_or_size is size */
+	if (data_offset_or_size < sizeof(struct ipc4_mixer_mode_config)) {
+		comp_err(dev, "mixin_set_large_config() too small data size: %u",
+			 data_offset_or_size);
+		return -EINVAL;
+	}
+
+	assert(data_offset_or_size <= SOF_IPC_MSG_MAX_SIZE);
```
Comment (lyakh) [inline src/audio/mixin_mixout.c:1140]: return an error too, I think

File: src/audio/mixin_mixout.c [hunk 6]
```diff
@@ -1020,6 +1111,73 @@ static int mixout_get_attribute(struct comp_dev *dev, uint32_t type, void *value
...
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)data, data_offset_or_size);
```
Comment (lyakh) [inline src/audio/mixin_mixout.c:1141]: is `data` a cached pointer? If it is, maybe the `.set_large_config` prototype should be changed to use `__sparse_cache`? In general - why is this needed? Is it shared between cores? Or accessed via cached and uncached aliases interchangeably?

Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:1141]: Large config payload comes from host via memory window. And so this invalidate cache is required. In fact, while developing in early version I did not have dcache_invalidate_region() call in my code and received garbage.

There are two possibilities, (1) to modify .set_large_config prototype to have __sparse_cache there, or (2) to call dcache_invalidate_region() internally before calling component's .set_large_config(). I think the second option might be nicer: I guess .set_large_config() is called from IPC task which always runs on core0.
But I'd suggest that this fix should be done as a separate PR and this PR could be merged as is.

File: src/audio/mixin_mixout.c (hunk 5 above)
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c:1140]: That's a check for data_offset_or_size should be less than IPC mailbox memory window size. In theory, it should never fail :) . In the worst case, if it fails, nothing terrible happen: data is casted to struct and used within boundaries of the struct. The worst thing is that dcache_invalidate_region() might be called with too large size.
So I'd rather remove that assert() at all, than replacing it with error report.

File: src/audio/mixin_mixout.c (hunk 4 above)
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Done.

File: src/audio/mixin_mixout.c [hunk 7]
```diff
@@ -1020,6 +1111,73 @@ static int mixout_get_attribute(struct comp_dev *dev, uint32_t type, void *value
...
+	}
+
+	/* for a single chunk data, data_offset_or_size is size */
+	if (data_offset_or_size < sizeof(struct ipc4_mixer_mode_config)) {
+		comp_err(dev, "mixin_set_large_config() too small data size: %u",
+			 data_offset_or_size);
+		return -EINVAL;
+	}
+
+	assert(data_offset_or_size <= SOF_IPC_MSG_MAX_SIZE);
```
[the remark this answers is not in the document]
Comment (serhiy-katsyuba-intel) [reply src/audio/mixin_mixout.c]: Done.

**Review by lgirdwood — APPROVED** [review]
